### DISCUSSION TITLE
Any chat such as Slack or Discord and such for ANTs and ANTs-derivative development?

### DISCUSSION THREAD

**Discussion by SirTwinkleberry** [ANTsX/ANTs#1643] [Q&A] [answered]
I'm in a the process of coding a GUI to facilitate the manipulation of our 3D and 4D volumes by interfacing with the relevant ANTs utilities.
I'm running into frequent issues, dumb questions or just need for more information, as well as issues linking the ANTs library C++ code to the main C++ GUI. I'm also running into issues when trying to build the master or the latest release on Windows using standard CMake configurations (which I'm bypassing by working on Linux).

All in all, I'm afraid to create an issue for something stupid and I was hoping there was a less formal channel of discussion available where I can ensure it's not just me being an idiot before opening an issue here?

**Comment by ntustison** [comment]
Sorry, we don't use Slack or Discord.  We didn't do a whole lot with GUI interfacing and there's not much Windows usage on the part of developers but we can try to answer whatever questions you might have.

**Comment by cookpa** [comment]
Discussions are threaded so feel free to post problems here. We can escalate to issues as needed. 

I'd not worry too much about it, people can mute this post if it gets too much for them

**Comment by SirTwinkleberry** [comment]
Thank you both for your prompt answer. While you were answering I was typing up a desperate post on Reddit ahah. I'll link it here for more context as well as summarize my problem below for longevity.
The post: https://www.reddit.com/r/cpp_questions/comments/18rb28d/another_post_about_external_libraries/

My problem:
I have started a Qt C++ project using the Qt Creator IDE in order to learn how to use their GUI / .ui files to create C++ UIs. On Windows, I have been using CMakeLists.txt to build my project, on Linux, I have been using a QMake .pro file because I was desperate most of the documentation online regarding Qt and external libraries use the older QMake.

Both on Windows and Linux I was able to build ANTs and its dependencies (ITKv5). On Linux, I installed ANTs in the default `/opt/ANTs/` directory where I can find both its executable and its static and dynamic lib files. On Windows, I kept everything in the `ANTS/build/` directory.

I'll continue with Linux only since this is the furthest I've come to make it work lol.
In my mainwindow.cpp file, I included the `PrintHeader.h` file as an example to try to make ANTs work with my code. Unfortunately, upon making and installing ANTs, no `include` directory is created, so to find the header file I add the path to the `PrintHeader.h` file within the source code I cloned from git (`INCLUDEPATH += ANTs/Examples/include`). I add the `PrintHeader` library to my project's make file using the LIBS variable (`LIBS += -L"/opt/ANTs/lib" -ll_PrintHeader`). Before doing that (when I first posted this message here), I was wrestling with `undefined reference to ants::PrintHeader`, now I'm wrestling with ~535 undefined reference errors, which all seem to stem from ITK libraries (for example: `:-1: error: /opt/ANTs/lib/libITKIONIFTI-5.4.so.1: undefined reference to is_nifti_file'`).

I saw on your closed issues that I was not the only one to have trouble linked ANTs to their project, but it seems that they all got it to work at some point and they didn't have to include hundreds of libraries either. So I'm thinking I'm doing something terribly wrong and I'd love a (std::shared_) pointer in the right direction please!


attached: the errors and warnings output file after trying to compile the project using the `PrintHeader` library.
[warningsAnderrorsList.txt](https://github.com/ANTsX/ANTs/files/13772655/warningsAnderrorsList.txt)

**Comment by SirTwinkleberry** [comment, marked as answer]
Okay so, in all my stupidity, I don't understand why but somehow I fixed it a bit.

I've added the line `-Wl,-rpath-link,"/opt/ANTs/lib` to the LIBS variable to try to get rid of the warning of missing dynamically linked libraries.
```bash
INCLUDEPATH += \
    /home/tanderson/ANTs/Examples/include \

LIBS += \
    -L"/opt/ANTs/lib" -ll_PrintHeader \
    -Wl,-rpath-link,"/opt/ANTs/lib"
```

This compiles my source and uses the `ants::PrintHeader` function just right. Perfect I guess? But, does that mean my code can only work with the available ANTs installation on the target computer, to dynamically link upon executing the code? How can I fully statically link everything? I should use the `-L"/opt/ANTs/lib"` command with an option to link every library as needed instead of just `-ll_PrintHeader`? Is that possible? Or do I have to link them all by hand?

I'll just try to understand this and then close the discussion at the risk of opening another one later if I have another burning question.

Hopefully I can get a GUI prototype to work soon, and maybe if I make it nice enough you may be interested. ^^

**Reply by SirTwinkleberry** [reply]
Doing it by hand, my `LIBS` variable looks like:
```
LIBS += \
    -L"/opt/ANTs/lib" \
    -ll_PrintHeader \
    -litkhdf5_cpp-shared-5.4 \
    -litkhdf5-shared-5.4 \
    -litkopenjpeg-5.4 \
    -litkminc2-5.4 \
    -lITKIOIPL-5.4 \
    -litkgdcmMSFF-5.4 \
    -litkgdcmDICT-5.4 \
    -litkgdcmDSED-5.4 \
    -litkzlib-5.4 \
    -litkjpeg-5.4 \
    -lITKMetaIO-5.4 \
    -lITKniftiio-5.4 \
    -lITKNrrdIO-5.4 \
    -litkpng-5.4 \
    -litkdouble-conversion-5.4 \
    -litkv3p_netlib-5.4 \
    -litktiff-5.4 \
    -litkgdcmjpeg8-5.4 \
    -litkgdcmjpeg12-5.4 \
    -litkgdcmjpeg16-5.4 \
    -litkgdcmopenjp2-5.4 \
    -litkgdcmcharls-5.4 \
    -litkgdcmuuid-5.4 \
    -litkgdcmIOD-5.4 \
    -litkgdcmCommon-5.4 \
    -lITKEXPAT-5.4
```

unfortunately, I checked the executable and it turns out they're all dynamically linked lol... Even `PrintHeader`... But that's less of an ANTs issue and more of a Me not using the make file properly I guess so it's all good. :)
